File: examples/python/async_device.py
#!/usr/bin/env python3
"""Example of asynchronous SHDLC device communication using asyncio."""

import asyncio
from rust_shdlc_driver import AsyncShdlcConnection, AsyncShdlcDevice, ShdlcSerialPort


async def main():
    # Connect to device on /dev/ttyUSB0 with baudrate 115200
    port = ShdlcSerialPort(port="/dev/ttyUSB0", baudrate=115200)
    conn = AsyncShdlcConnection(port)
    device = AsyncShdlcDevice(conn, slave_address=0)

    product_name = await device.get_product_name()
    product_type = await device.get_product_type()
    serial_number = await device.get_serial_number()
    version = await device.get_version()

    print(f"Product Name: {product_name}")
    print(f"Product Type: {product_type}")
    print(f"Serial Number: {serial_number}")
    print(f"Version: {version}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/async_device.rs
use rust_shdlc_driver::connection::AsyncShdlcConnection;
use rust_shdlc_driver::device::AsyncShdlcDevice;
use rust_shdlc_driver::transport::AsyncSerialPort;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Open serial port at 115200 baud
    let mut port = AsyncSerialPort::new("/dev/ttyUSB0", 115200);
    port.open()?;

    let conn = AsyncShdlcConnection::new(Box::new(port));
    let mut device = AsyncShdlcDevice::new(conn, 0);

    // Read device information asynchronously
    let product_name = device.get_product_name().await?;
    let product_type = device.get_product_type().await?;
    let serial_number = device.get_serial_number().await?;
    let version = device.get_version().await?;

    println!("Product Name: {}", product_name);
    println!("Product Type: {}", product_type);
    println!("Serial Number: {}", serial_number);
    println!(
        "Firmware Version: {}.{}",
        version.firmware.major, version.firmware.minor
    );

    Ok(())
}
